Add IsCloseTo check for DateTime and DateTimeOffset

ocoanet opened the pull request:
Fixes #312.

dupdob commented:
Many thanks for this PR. While I think its purpose is self explanatory, could you please open an issue to describe the (expected) behaviour for your new check?

dupdob commented on code/src/NFluent/Messages/DateTimeRangeBlock.cs:
```diff
@@ -0,0 +1,66 @@
+    internal class DateTimeRangeBlock
+    {
+        private readonly DateTime referenceValue;
+        private readonly TimeSpan tolerance;
```
This should be a Duration instead of a TimeSpan: the Duration class has been designed so that the preferred time unit is kept to ensure a easy to read output!

ocoanet replied:
The value of `Duration` seems to be the nice `ToString` and the easy to read constructors. That's why I provided an `IsCloseTo` override with `Duration`. But the range only need the timespan for the `IsInRange` check.

ocoanet replied:
The `toleranceString` is here specifically to preserve the easy to read output from `Duration`.

dupdob commented on code/src/NFluent/Messages/DateTimeRangeBlock.cs:
```diff
@@ -0,0 +1,66 @@
+        /// <param name="referenceValue">Reference value (mid point)</param>
+        /// <param name="tolerance">Tolerance</param>
+        /// <remarks>This represents a range of <see cref="referenceValue" /> +/- <see cref="tolerance" /></remarks>
+        public DateTimeRangeBlock(DateTime referenceValue, TimeSpan tolerance)
```
This constructor should not exist (see comment about Duration vs TImeSpan

ocoanet replied:
I can remove this constructor and generate a `Duration` from the `TimeSpan`. Is that ok for you?

dupdob replied:
That's the idea. The next question is : how to select an appropriate time unit when none is provided. You may add a (static) method to the Duration class to identity the appropriate unit

dupdob commented on code/tests/NFluent.Tests/DateTimeOffsetRelatedTests.cs:
```diff
@@ -411,5 +412,55 @@ public void NotMatchTheSameUtcInstantWorks()
+            var within = TimeSpan.FromMilliseconds(1);
+
+            Check.ThatCode(() => Check.That(reference).IsCloseTo(expected, within))
+                 .IsAFailingCheckWithMessage("",
```
You should use a constant date so that the error message is a constant! In this case, it would highlight that TimeSpan does not provide a simple (to read) output.

dupdob commented on code/tests/NFluent.Tests/DateTimeRelatedTests.cs:
```diff
@@ -957,6 +958,56 @@ public void AndOperatorCanChainMultipleAssertionOnDateTime()
+            var expected = reference.AddMilliseconds(2);
+            var within = TimeSpan.FromMilliseconds(1);
+
+            Check.ThatCode(() => Check.That(reference).IsCloseTo(expected, within))
```
You should use a constant date so that the error message is a constant! In this case, it would highlight that TimeSpan does not provide a simple (to read) output.

ocoanet replied:
I think that Duration is nice but I still would like to provide a TimeSpan overload. Many users might be already using extensions methods like FluentDateTime to create TimeSpan.

dupdob commented on code/tests/NFluent.Tests/DateTimeRelatedTests.cs:
```diff
@@ -957,6 +958,56 @@ public void AndOperatorCanChainMultipleAssertionOnDateTime()
+            var within = new Duration(1, TimeUnit.Milliseconds);
+
+            Check.ThatCode(() => Check.That(reference).IsCloseTo(expected, within))
+                 .IsAFailingCheckWithMessage("",
```
You should use a constant date so that the error message is a constant! In this case, it would highlight that TimeSpan does not provide a simple (to read) output.

dupdob commented on code/tests/NFluent.Tests/DateTimeOffsetRelatedTests.cs:
```diff
@@ -411,5 +412,55 @@ public void NotMatchTheSameUtcInstantWorks()
+            var within = new Duration(1, TimeUnit.Milliseconds);
+
+            Check.ThatCode(() => Check.That(reference).IsCloseTo(expected, within))
+                 .IsAFailingCheckWithMessage("",
```
You should use a constant date so that the error message is a constant! In this case, it would highlight that TimeSpan does not provide a simple (to read) output.

dupdob commented on code/src/NFluent/Messages/DateTimeOffsetRangeBlock.cs:
```diff
@@ -0,0 +1,66 @@
+    internal class DateTimeOffsetRangeBlock
+    {
+        private readonly DateTimeOffset referenceValue;
+        private readonly TimeSpan tolerance;
```
This should be a Duration instead of a TimeSpan: the Duration class has been designed so that the preferred time unit is kept to ensure a easy to read output!

dupdob commented on code/src/NFluent/Messages/DateTimeOffsetRangeBlock.cs:
```diff
@@ -0,0 +1,66 @@
+        /// <param name="referenceValue">Reference value (mid point)</param>
+        /// <param name="tolerance">Tolerance</param>
+        /// <remarks>This represents a range of <see cref="referenceValue" /> +/- <see cref="tolerance" /></remarks>
+        public DateTimeOffsetRangeBlock(DateTimeOffset referenceValue, TimeSpan tolerance)
```
This constructor should not exist (see comment about Duration vs TImeSpan

ocoanet commented on code/src/NFluent/Messages/DateTimeOffsetRangeBlock.cs:
```diff
@@ -0,0 +1,66 @@
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace NFluent.Messages
```
I put it here because `RangeBlock` was here too. I wanted to follow the same pattern.

dupdob replied:
Good thinking. But my conclusions is that I need to move Rangeblock as well.
I do it. Thanks for helping me spot this design smell.
The messages namespace should definitely only contains error message oriented classes.

ocoanet commented on code/src/NFluent/Messages/DateTimeRangeBlock.cs:
```diff
@@ -0,0 +1,66 @@
+// </copyright>
+// --------------------------------------------------------------------------------------------------------------------
+
+namespace NFluent.Messages
```
I put it here because `RangeBlock` was here too. I wanted to follow the same pattern.

dupdob replied:
Good thinking. But my conclusions is that I need to move Rangeblock as well.
I do it. Thanks for helping me spot this design smell.
The messages namespace should definitely only contains error message oriented classes.

ocoanet commented:
Is that ok with you if I keep the TimeSpan based overloads but then converts the TimeSpan to a Duration?

ocoanet commented:
I just updated the PR:
- The block types are in now the Helpers namespace.
- The block types have a `Duration` field instead of TimeSpan.
- The `Duration.ToTimeSpan` was no longer necessary and was removed.
- I added a new factory method for Duration: `Duration.FromTimeSpan`. It is not required, I can remove it or rename it if needed.
- The code is rebased after your latest changes on master.

I used `TimeHelper.DiscoverUnit` and I was surprised to discover that the `MinimumUnitAmount` is 2 instead of 1. In my view _"1 Seconds"_ is easier to read than _"1000 Milliseconds"_. I did not change this behavior though.

Tell me if anything needs to be corrected or if the changes need to be squashed.

dupdob commented:
Many thanks for these adjustments!
> I used TimeHelper.DiscoverUnit and I was surprised to discover that the MinimumUnitAmount is 2 instead of 1. In my view "1 Seconds" is easier to read than "1000 Milliseconds". I did not change this behavior though.

Thanks, here is the rationale: duration are expressed in integer time units, hence the choice of 2 has a minimal value to avoid rounding 1499 ms to 2. Not a very strong case, so I am open to other views.

Just one last request: could you add a test case for Not, especially for when **Not.IsCloseTo** fails, to have an assertion on the test message?
I am afraid I forgot to ask for it in my initial review.

I will update the contributing.md file to list the requested tests.

ocoanet commented:
> Just one last request: could you add a test case for Not, especially for when Not.IsCloseTo fails, to have an assertion on the test message?

Done.

dupdob commented:
Great additions!

dupdob commented:
Thanks for a great conribution.
